### ISSUE TITLE
ResourceConflictException on first run of each new input configuration

### ISSUE THREAD

**Issue by stevehogdahl** [alexcasalboni/aws-lambda-power-tuning#124]
The first time i run a new configuration for a lambda i get the following error. I thought it might have to do with the aliasing mentioned in the deployment instructions, but my input contains the autoOptimizeAlias option. Once it fails the first time, i rerun it using the exact same input and it's successful. Is this expected or do you know how to solved this? Thank you

`2020-12-23T01:09:32.483Z	d9613fd9-1ec5-4874-b72f-4a30ecc7ed96	ERROR	ResourceConflictException: Unable to delete version because the following aliases reference it: [latest]    at Object.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:52:27)    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/rest_json.js:55:8)    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:688:14)    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:690:12) {  code: 'ResourceConflictException',  time: 2020-12-23T01:09:32.404Z,  requestId: '9dc41156-3376-41f6-aa1a-0994f3c6bcd4',  statusCode: 409,  retryable: false,  retryDelay: 42.32576658491525}`

**Comment by alexcasalboni** [comment]
Thanks for reporting, @stevehogdahl.

This might happen if the state machine execution fails or if you try to execute it multiple times in parallel.

A couple of things that might help debugging:

1. Does it happen even without the `autoOptimizeAlias` option?
2. Could you please share your state machine input? (feel free to omit or anonymize the payload)

**Comment by stevehogdahl** [comment]
Sorry for taking a while to get back to you on this. Had to put it down, but i'll get back to you with an update and more details in the next week or so. thank you.

**Comment by j0v3** [comment]
This might be also related.
I'm running the tool over a couple of lambda@edge functions, and the first time I run them on a new state machine it tries to delete my actual lambdas (and not just the aliased ones):
```json
  {
    "lambdaFunctionFailedEventDetails": {
        "error": "InvalidParameterValueException",
        "cause": {
            "errorType": "InvalidParameterValueException",
            "errorMessage": "Lambda was unable to delete arn:aws:lambda:us-east-1:MY_AWS_ACCOUNT:function:MY_LAMBDA:58 because it is a replicated function. Please see our documentation for Deleting Lambda@Edge Functions and Replicas.",
            "trace": [
                "InvalidParameterValueException: Lambda was unable to delete arn:aws:lambda:us-east-1:MY_AWS_ACCOUNT:function:MY_LAMBDA:58 because it is a replicated function. Please see our documentation for Deleting Lambda@Edge Functions and Replicas.",
                "at Object.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:52:27)",
                "at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/rest_json.js:55:8)",
                "at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)",
                "at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)",
                "at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:688:14)",
                "at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)",
                "at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)",
                "at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10",
                "at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)",
                "at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:690:12)"
            ]
        }
    }
}
```
Just to add more context: I'm not using `parallelInvocation` execution or `autoOptimizeAlias`

One of the lambda versions that is getting targeted on clean up step matches the version that my actual lambda has, hence, the run fails. On subsequent runs of the tool, it creates aliased lambdas that have higher version numbers and they don't match the current lambda version (unless you deploy a new lambda version when the problem looks to show again).

Below you can see that the tool has even aliased the lambda version `#58` which is the version of my current replicated lambda. 
![Screenshot 2021-03-03 at 12 18 22](https://user-images.githubusercontent.com/38042646/109805389-48118100-7c1b-11eb-9375-4ea357bb513a.png)

**Comment by alexcasalboni** [comment]
Thanks for sharing @j0v3! I think this might be related to L@E specifically, as I've never tried to use Lambda Power Tuning with edge functions. I'll create a new issue to dive deeper into L@E support!

Looking at the error, the ARN is `arn:aws:lambda:us-east-1:MY_AWS_ACCOUNT:function:MY_LAMBDA:58` (where `58` is the version) - which is expected. Lambda Power Tuning creates (and then deletes) new versions and aliases to invoke all of them in parallel. Even if you disable `parallelInvocation`, the different configurations run in parallel.

@stevehogdahl I believe your error was different. You were not using L@E functions, right? Could you share more details?

**Comment by alexcasalboni** [comment]
@stevehogdahl I haven't been able to reproduce this error yet. Could you please share some additional details?

**Comment by stevehogdahl** [comment]
@alexcasalboni I've stopped using the library, you can close the issue if you're unable to reproduce.

**Comment by alexcasalboni** [comment]
No problem, thanks @stevehogdahl!

I'll close this for now.
